src/cmdstan/command_helper.hpp

```diff
@@ -766,8 +766,9 @@ void init_callbacks(
         &sample_writers,
     std::vector<stan::callbacks::unique_stream_writer<std::ofstream>>
         &diag_csv_writers,
+    std::vector<stan::callbacks::json_writer<std::ofstream>> &diag_json_writers,
     std::vector<stan::callbacks::json_writer<std::ofstream>>
-        &diag_json_writers) {
+        &metric_json_writers) {
```

Comment:
I think a fourth in-out parameter on this function is probably where we should draw the line and split it up or refactor it.

I think you can simplify this logic a lot by writing two helper functions:

```c++
template <typename T>
void null_initialize_callbacks(std::vector<T> &writers, size_t num_chains) {
  writers.reserve(num_chains);
  for (size_t i = 0; i < num_chains; ++i) {
    writers.emplace_back(nullptr);
  }
}

template <typename T, typename... Ts>
void file_initialize_writers(std::vector<T> &writers, unsigned int num_chains,
                             unsigned int id, std::string &filename,
                             std::string tag, std::string suffix, int sig_figs,
                             Ts &&...args) {
  writers.reserve(num_chains);
  auto filenames = make_filenames(filename, tag, suffix, num_chains, id);

  for (size_t i = 0; i < num_chains; ++i) {
    auto ofs = std::make_unique<std::ofstream>(filenames[i]);
    if (sig_figs > -1) {
      ofs->precision(sig_figs);
    }
    writers.emplace_back(std::move(ofs), std::forward<Ts>(args)...);
  }
}
```

The first will fill the callback lists with nullptrs, the second will create files. You could use them in the existing function like:

```
  // default - no diagnostics
  null_initialize_callbacks(diag_csv_writers, num_chains);
  null_initialize_callbacks(diag_json_writers, num_chains);
```

```
else {  // not pathfinder

    file_initialize_writers(sample_writers, num_chains, id, output_file, "",
                            ".csv", sig_figs, "# ");

```

But I think it might be better to break this function apart and inline parts of it using those helpers.

Reply:
changed per your suggestion.  many thanks!